### ISSUE TITLE
Schema actions do not populate StaticFields list

### ISSUE THREAD

**Issue by jimitndiaye** [Parature/ParatureSDK#30]
This is based on v2.0.8. 

When calling the Schema action the static fields of the returned entity are not populated, requiring the getter to be accessed for each individual property of the entity or to parse the *XmlReceived* property of the ApiCallResponse (I've resorted to this). It is important for my particular integration that the full structure of entity (including all fields, their datatypes etc) be available upon request without resorting to something like .net reflection.

From inspection of the code it seems the static fields retrieved from the server xml are deliberately stripped out to avoid deserialization problems. To work around this I've used xml queries on the *XmlReceived* to determine the available static fields and their properties. It would be great if this was built in.

**Comment by brtubb** [comment]
What specific functionality are you looking for? If I understand you correctly, you are looking to retrieve the list of Static Fields at run-time? Compile time definitely isn't possible without a code-gen tool (which we likely won't make). The reason the Static Field list is stripped is to limit the number of Nullable properties on the classes. The XmlSerializer will choke on deserializing the Schema XML due to the null nodes and non-null class properties.

Best I foresee us adding would be a property on the ParaEntity base class -> a list of the Static Fields returned from the Schema call. This would not be a particularly elegant solution, however.

**Comment by jimitndiaye** [comment]
Yes, we'd like to be able to determine the structure of an entity at run-time. This would include such information as the data type, whether it is editable, required, etc. If you're familiar with Dynamics CRM for instance, you can retrieve entity metadata that describes the structure of the entity. I expected that the Schema action would do the same thing returning an entity with all its staticfields and customfield descriptors populated. It currently does that for custom fields but not static fields. At runtime you don't always know the type of entity you're dealing with (other than it is a *ParaEntity*) and having access to the metadata enables us to deal with all entities in flexible and deterministic manner. Otherwise you end up with a lot of switch-case or if-else statements.

**Comment by jimitndiaye** [comment]
Specifically, if I make a call like this:

var accountSchema = Account.Schema(credentials);

Then accountSchema.StaticFields does not have the fully populated list of static fields available on the entity.

**Comment by brtubb** [comment]
I like the suggestion, and is definitely feasible. The deserialization had to strip this info to properly work, but the info is at least retrievable. We'd likely have to put this information in a separate property due to static fields not all being Nullable, but we'll try to not do that.

**Comment by jimitndiaye** [comment]
Excellent. I look forward to your solution. I have related question specifically regarding date StaticFields. I've been using the FieldType attribute to differentiate between DateTime values that are meant to be date-only fields and those for which the time is significant. I assumed (perhaps wrongly) that fields with DataType=date and FieldType=usdate were date-only fields (e.g 1/10/2015) and those for which FieldType is some other value are date-time fields (e.g. 1/10/2015 09:52 UTC). The difference between the two is obviously important in order to cater for time zone differences and display formatting. For instance you don't do time-zone conversion on DateTime values that are meant to be date-only.

Giving Dynamics CRM as an example again, they have an attribute called Format of type DateTimeFormat on DateTimeAttributeMetadata instances which determines which of the two representations the field models.

**Comment by brtubb** [comment]
To simplify the matter of Date vs DateTime fields - only "Static" fields are DateTime. Custom Fields are all Dates only, and do not have a time component.

No doubt CRM has a better schema representation in their API/SDKs. Our SDK strives to simulate the CRM SDK in many ways, but this is an area we can't get 100% parity.

**Comment by jimitndiaye** [comment]
This is still an issue in verion 2.2.2. Additionally since the ```FirsLevelApiGetMethods.Schema``` function is being deprecated it makes matters worse for us since we'll no longer get access to underlying XML to retrieve the missing data.
Fundamentally, we need access not only to the list of static fields for each entity, but to each fields metadata such as name, data type, required, editable, field-type, maxLength, etc. This is provided when you use ParaService.Create because again the static fields are stripped from the xml. If you're concerned about the xml serializer stripping over some properties then you can extract the necessary information on a field by field basis. For instance, this is what we do currently to get all the static field metadata: 
```C#
                    var xml = XDocument.Parse(schema.ApiCallResponse.XmlReceived.OuterXml);
                    if (xml.Root != null)
                    {
                        bool isEditable, isRequired;
                        int maxLengthValue;
                        var staticFields = from node in xml.Root.Elements()
                            where node.Name.LocalName != "Custom_Field"
                            let fieldName = node.Name.LocalName
                            //where !schema.StaticFields.Exists(f => f.Name == fieldName)
                            let requiredAttribute = node.Attribute("required")
                            let required = requiredAttribute != null &&
                                           bool.TryParse(requiredAttribute.Value, out isRequired) &&
                                           isRequired
                            let editableAttribute = node.Attribute("editable")
                            let editable = editableAttribute != null &&
                                           bool.TryParse(editableAttribute.Value, out isEditable) &&
                                           isEditable
                            let fieldTypeAttribute = node.Attribute("field-type")
                            let fieldType = fieldTypeAttribute != null ? fieldTypeAttribute.Value : null
                            let dataTypeAttribute = node.Attribute("data-type")
                            let dataType = dataTypeAttribute != null ? dataTypeAttribute.Value : null
                            let maxLengthAttribute = node.Attribute("maxLength")
                            let maxLength = maxLengthAttribute != null &&
                                            int.TryParse(maxLengthAttribute.Value, out maxLengthValue)
                                ? maxLengthValue
                                : 0
                            select new StaticField
                                   {
                                       Name = fieldName,
                                       DataType = dataType,
                                       Editable = editable,
                                       Required = required,
                                       FieldType = fieldType,
                                       MaxLength = maxLength
                                   };
                        foreach (var field in staticFields)
                            schema[field.Name] = field;
                        //schema.StaticFields.AddRange(staticFields);
                    }
```
If there is another way to retrieve these StaticField properties, please advise.
